### PR TITLE
[feature] Clear management_ip when device goes offline #286

### LINKED ISSUE #286: [change] Remove management_ip from device when status changes to offline
We need to introduce a setting, set to `True` by default, called `OPENWISP_MONITORING_AUTO_CLEAR_MANAGEMENT_IP`, which when `True` automatically clears the `management_ip` of a device when its monitoring status passes to `critical`.
We can do this directly in `AbstractDeviceMonitoring.update_status()`.

Reason: whenever a single OpenVPN instance or pool of instances using the same DHCP server is shared among different organizations, the same IP can be assigned first to the device of org A, then to the device of org B, but the management IP saved in device of org A is not cleared so we end up with two devices having the same management IP.
OpenWISP controller already does this cleanup, but it only works within the same organization, different organizations can have the same management IP because they could use different VPN instances.

### PR DESCRIPTION
**Pull request by devkapilbansal**
Closes #286

Checks:

- [x] I have manually tested the proposed changes
- [x] I have written new test cases to avoid regressions (if necessary)
- [x] I have updated the documentation (e.g. README.rst)

### CODE REVIEW

File: openwisp_monitoring/device/base/models.py [hunk 1]
```diff
@@ -253,6 +253,11 @@ def update_status(self, value):
         self.status = value
         self.full_clean()
         self.save()
+        # clear device management_ip when device is offline
+        if self.status == 'critical' and app_settings.AUTO_CLEAR_MANAGEMENT_IP:
+            self.device.management_ip = None
+            self.device.save()
```
Comment (nemesisdesign) [inline openwisp_monitoring/device/base/models.py:259]: Try:

`self.device.save(update_fields=['management_ip'])`

It's more efficient.

File: tests/openwisp2/settings.py [hunk 2]
```diff
@@ -144,6 +144,7 @@
 EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
 
 OPENWISP_MONITORING_MANAGEMENT_IP_ONLY = False
+OPENWISP_MONITORING_AUTO_CLEAR_MANAGEMENT_IP = True
```
Comment (nemesisdesign) [inline tests/openwisp2/settings.py:147]: This line should not be needed because it's the default value, or what am I missing?

File: README.rst [hunk 3]
```diff
@@ -804,6 +804,19 @@ In case you just want to change the colors used in a chart here's how to do it:
...
++--------------+-------------+
+| **type**:    | ``bool``    |
++--------------+-------------+
+| **default**: | ``True``    |
++--------------+-------------+
+
+This setting allows you to automatically clear management_ip of a device
+when it goes offline. It is enabled by default.
+
+
```
Comment (nemesisdesign) [inline README.rst:819]: please avoid this extra blank line

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Looks good @devkapilbansal, see my comments below.

PS: also change the prefix of your commit to `[feature]` please.

File: tests/openwisp2/settings.py (hunk 2 above)
Comment (devkapilbansal) [reply tests/openwisp2/settings.py]: You're right. I should remove this

File: openwisp_monitoring/device/base/models.py (hunk 1 above)
Comment (devkapilbansal) [reply openwisp_monitoring/device/base/models.py]: I was unaware of this. Thanks :smile:

File: openwisp_monitoring/device/tests/test_models.py [hunk 4]
```diff
@@ -590,6 +590,16 @@ def test_ok_problem_problem_problem_ok(self):
         load.check_threshold(80)
         self.assertEqual(dm.status, 'ok')
 
+    def test_management_ip_clear_device_offline(self):
```
Comment (devkapilbansal) [inline openwisp_monitoring/device/tests/test_models.py:593]: Should there be a test case too to ensure management_ip is not cleared when `AUTO_CLEAR_MANAGEMENT_IP` is false?

Comment (pandafy) [reply openwisp_monitoring/device/tests/test_models.py:593]: Yes, that would be great to have.

**Review by nemesisdesign — APPROVED** [review]
Very good work @devkapilbansal!

I changed a minor detail in the tests: https://github.com/openwisp/openwisp-monitoring/pull/292/commits/7f7b25f8e393dcccdee70e3e97d1986b30024d86
Please always ensure the data in the test is somewhat realistic, `0.0.0.0` does not make sense as a `management_ip` and even if for the sake of the test it doesn't change whether the value is realistic or not, it does affect the readability of the code, moreover, in the eye of a potential user who may be a network engineer who is inspecting our code, it can make a difference in the way they judge our work.

**Pull request merged**